Pass on embedded layout section (#752)
browserengineering/book: book/embeds.md

Embedded layout
===============

> By now, you probably have a good sense of how to add images to our

Reviewer: ```suggestion
Based on your experience with prior chapters, you can probably guess how to add images to our
```

potentially an entire run of text, split across multiple lines, while
an image is an [atomic inline][atomic-inline]. The other types of
embedded content in this chapter are also atomic inlines.

[atomic-inline]: https://drafts.csswg.org/css-display-3/#atomic-inline

Let's split the existing `InputLayout` into a superclass called
`EmbedLayout`, containing most of the existing code, and a new
subclass with the input-specific code, `InputLayout`:[^widgets]

[^widgets]: In a real browser, input elements are usually called
*widgets* because they have a lot of [special rendering
rules][widget-rendering] that sometimes involve CSS.

[widget-rendering]: https://html.spec.whatwg.org/multipage/rendering.html#widgets

``` {.python replace=previous/previous%2c%20frame}
class InputLayout(EmbedLayout):
    def __init__(self, node, parent, previous):
        super().__init__(node, parent, previous)

    def layout(self, zoom):
        super().layout(zoom)
```

Now, the idea is that `EmbedLayout` should provide common layout code
for all kinds of embedded content, while its subclasses like
`InputLayout` should provide the custom code needed to draw that specific

> kind of embedded content. Different types of embedded content might

Reviewer: ```suggestion
kind. Different types of embedded content might
```

different widths and heights, so that should happen in `InputLayout`.

> Likewise, `paint` should happen in `InputLayout`:

Reviewer: ```suggestion
Likewise, `paint` should happen there too:
```

real browser, it can be avoided by forcing an image into a block or
other layout mode via the `display` CSS property.] The underlying
reason for this is because, as a type of inline layout, images are
designed to flow along with related text, including the computation of
[baselines][baseline-ch3]. So a font is involved somehow. For example,
the baseline of the image should line up with the of the text next to
it.

[baseline-ch3]: text.html#text-of-different-sizes

Painting an image is also straightforward:

``` {.python}
class ImageLayout(EmbedLayout):
    def paint(self, display_list):
        cmds = []
        rect = skia.Rect.MakeLTRB(
            self.x, self.y + self.height - self.img_height,
            self.x + self.width, self.y + self.height)
        quality = self.node.style.get("image-rendering", "auto")
        cmds.append(DrawImage(self.node.image, rect, quality))
        display_list.extend(cmds)
```

Note that the saved `img_height` property is used to make sure the
image is positioned with its bottom edge on the text baseline.

Now we just need to hook up the new `ImageLayout` to `BlockLayout`. We
could do this by duplicating the `input` method and calling it

> `image`... but `input` is itself a duplicate of `text`, and after

Reviewer: ```suggestion
`image`...but `input` is itself a duplicate of `text`, and after
```
